### PR TITLE
MSDK-329: in-app SDK debug log overlay for Bonni

### PR DESCRIPTION
**Pull request by usharif-attentive**
## PR Type
- [x] Feature

## What's new?
Final piece of [MSDK-329](https://attentivemobile.atlassian.net/browse/MSDK-329) — an always-available, in-app SDK log viewer for the Bonni example app. Lets QA inspect SDK log output on TestFlight/release builds without attaching a debugger.

### SDK
- Wraps \`os.Logger\` with a new \`ATTNLogger\` struct that fans out to both the OS unified log AND an in-process ring buffer (\`ATTNLogBuffer\`). Existing 178 \`Loggers.creative.debug(...)\` / etc. call sites compile unchanged thanks to a custom \`ATTNLogMessage: ExpressibleByStringInterpolation\` that mirrors \`os.Logger\`'s \`, privacy:\` API.
- \`ATTNLogBuffer\` is a thread-safe ring buffer (1000 entries, amortized O(1) eviction) with a snapshot accessor and a multi-subscriber \`AsyncStream\`.
- New public API: \`ATTNSDK.recentLogs(since:)\` and \`ATTNSDK.logStream\`. Both work in any build configuration — including TestFlight — because the buffer captures \`.debug\` directly instead of relying on \`OSLogStore\`, which only persists \`.notice\` and above by default.
- New public type: \`ATTNLogEntry\` (\`@objc\`-bridged snapshot of one log line with a shared \`formatted(timestamp:)\` helper).

### Bonni
- New \`DebugLogOverlay\`: a separate \`UIWindow\` at \`windowLevel = .alert + 1\` installed once from \`SceneDelegate\` so it floats above every screen.
- A draggable \"LOGS\" toggle button shows/hides a compact, draggable floating panel (320×240) that streams entries live via \`ATTNSDK.logStream\` and renders them with timestamp, category, level, and message.
- The overlay window uses a \`hitTest\` override so non-chrome regions pass touches through to the underlying app — the rest of Bonni stays fully usable with the overlay installed.
- Panel work is skipped while hidden (entries queue to a small pending buffer and flush in one batch on toggle); visible appends use \`textStorage.append\` rather than reassigning \`textView.text\`, so per-entry cost stays O(1).

### Limitations / non-goals
- Privacy: every \`Loggers.*\` call is forwarded to \`os.Logger\` as a single \`.public\` value. Every existing call site already annotates \`.public\`, so this is a no-op today; documented inline so future contributors don't expect \`.private\` to redact in Console.app.
- Subscriber bookkeeping uses a UUID-keyed dict rather than reusing an extracted helper — there is no other \`AsyncStream\` subscriber-management code in the SDK to share with yet.

## Testing
- Builds the SDK framework and the Bonni app cleanly.
- Verified in the iOS Simulator: the floating LOGS button overlays the login screen with login UI fully tappable. Real SDK entries (\`network DEBUG: Success on attempt 1\`, \`event INFO: Obtained existing visitor id: ...\`, \`event DEBUG: Successfully sent push token\`) stream into the panel live.

## Screenshots
https://github.com/user-attachments/assets/bad66b06-cdbf-4191-8024-17c1cf6f9185

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[MSDK-329]: https://attentivemobile.atlassian.net/browse/MSDK-329?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Sources/Helpers/ATTNLogBuffer.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            guard let since else { return entries }
+            return entries.filter { $0.date >= since }
+        }
+    }
+
+    /// Async stream of new entries appended after subscription. The caller must
+    /// retain the resulting Task; cancelling it removes the subscription.
+    func stream() -> AsyncStream<ATTNLogEntry> {
+        AsyncStream { continuation in
+            let id = UUID()
```
Comment (adelag-attentive) [inline Sources/Helpers/ATTNLogBuffer.swift:68]: `AsyncStream` without an explicit buffering policy defaults to unbounded buffering. If a subscriber consumes slower than logs are produced (UI hidden, Task suspended), yielded entries accumulate in memory without limit. Since `ATTNSDK.logStream` is public and multi-subscriber, one lagging consumer can grow unboundedly while others keep up.

Recommend specifying a bounded policy:
```swift
AsyncStream(bufferingPolicy: .bufferingNewest(100)) { continuation in ... }
```
This drops the oldest pending entries when a consumer falls behind — acceptable for a debug tool.

File: Sources/Public/ATTNLogEntry.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+@objc(ATTNLogEntry)
+public final class ATTNLogEntry: NSObject {
+    public enum Level: String, CaseIterable {
+        case debug
+        case info
+        case notice
+        case warning
+        case error
+        case fault
+    }
```
Comment (adelag-attentive) [inline Sources/Public/ATTNLogEntry.swift:20]: `Level` is defined as a typed enum, but the `level` property is stored as a raw `String`. Swift consumers lose compiler-checked exhaustiveness and must do manual string matching.

Presumably this is for `@objc` bridging. Consider adding a computed property for Swift callers:
```swift
public var logLevel: Level? { Level(rawValue: level) }
```
At minimum, document that `level` always contains one of the `Level.rawValue` strings.

File: Sources/Helpers/Logger.swift [hunk 3]
```diff
@@ -8,10 +8,100 @@
...
+
+    func info(_ message: ATTNLogMessage) {
+        osLogger.info("\(message.rendered, privacy: .public)")
+        ATTNLogBuffer.shared.append(level: .info, category: category, message: message.rendered)
+    }
+
+    func notice(_ message: ATTNLogMessage) {
+        osLogger.notice("\(message.rendered, privacy: .public)")
+        ATTNLogBuffer.shared.append(level: .notice, category: category, message: message.rendered)
+    }
```
Comment (adelag-attentive) [inline Sources/Helpers/Logger.swift]: `os.Logger` defers string interpolation until the system decides to persist the message. On release builds, `.debug` entries are typically discarded before interpolation occurs — zero cost.

With `ATTNLogMessage`, interpolation is evaluated eagerly at every call site to populate `rendered`, regardless of whether `os.Logger` would have discarded it. On hot paths (e.g., debug logging on every network response), this overhead could be measurable.

Consider gating buffer appends behind a runtime flag (e.g., `ATTNLogBuffer.shared.isEnabled`) so builds that don't use the overlay pay nothing, or benchmark to confirm the cost is acceptable.

File: Bonni/AttentiveExample/DebugLogOverlay.swift [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    private func appendVisibleLine(_ line: String) {
+        let prefix = textView.text.isEmpty ? "" : "\n"
+        textView.textStorage.append(NSAttributedString(
+            string: prefix + line,
+            attributes: [
+                .font: textView.font ?? UIFont.monospacedSystemFont(ofSize: 10, weight: .regular),
+                .foregroundColor: textView.textColor ?? UIColor.white
+            ]
```
Comment (adelag-attentive) [inline Bonni/AttentiveExample/DebugLogOverlay.swift]: The panel uses `textStorage.append` for O(1) per-entry appends, but `trimIfNeeded()` walks the entire text from the start to locate newlines (O(n)), then reassigns `textView.text` — discarding the existing `textStorage` and rebuilding from scratch.

There is also a bookkeeping issue: `flushPending()` joins multiple entries with `"\n"` into a single string and calls `appendVisibleLine` once, incrementing `lineCount` by 1. The actual text contains multiple lines, so `lineCount` undercounts and trimming triggers later than expected.

Consider maintaining entries in a `[String]` array and rebuilding from the tail on trim, or using `textStorage.replaceCharacters(in:with:"")` to remove the leading range without full reassignment.

**Review by adelag-attentive — COMMENTED** [review]
General comment: The ring buffer has non-trivial eviction logic (amortized `capacity + evictionSlack` boundary) and the subscriber lifecycle has edge cases (e.g., `onTermination` firing concurrently with appends). Core SDK additions like this should have unit test coverage — particularly the eviction boundary and subscriber cleanup path.

File: Sources/Helpers/Logger.swift (hunk 3 above)
Comment (adelag-attentive) [inline Sources/Helpers/Logger.swift:72]: **Eager interpolation cost on release builds**

With `os.Logger`, `.debug` messages on release builds are essentially free — the system discards them *before* evaluating the string interpolation closure. The interpolation is never executed, so there's zero CPU or allocation cost at the call site.

`ATTNLogMessage` changes this contract. Because it conforms to `ExpressibleByStringInterpolation`, the compiler materializes the full `ATTNLogMessage` value (including evaluating every interpolation segment into `rendered`) *at the call site*, before `debug(_:)` is even entered. So by the time your function body runs and `ATTNLogBuffer.shared.append()` checks `isCapturing`, the string has already been built.

Concretely, a call like:
```swift
Loggers.creative.debug("Resizing creative to \(newArea.width)x\(newArea.height)")
```
…previously cost **nothing** on a release build (os.Logger discards `.debug` without interpolating). Now it unconditionally allocates a String and runs `String.append` for each segment — even when `isCapturing = false` and the message is thrown away.

For infrequent log sites this doesn't matter, but the SDK has debug logging on network callbacks, creative resize events, and navigation decisions — paths that can fire many times per second.

**Options:**
1. Accept the cost and document it (if benchmarking shows it's negligible for our call sites).
2. Make `debug(_:)` accept `@autoclosure () -> ATTNLogMessage` so interpolation is deferred until you confirm the buffer wants it. You'd still forward to `os.Logger` with its native interpolation separately (preserving its lazy semantics), and only evaluate the autoclosure when `isCapturing` is true.
3. Use a compile-time flag to strip buffer appends entirely from production builds that don't ship the overlay.

I'd lean toward at least benchmarking before merging to confirm the overhead is acceptable for the hot paths.

File: Sources/Public/SDK/ATTNSDK.swift [hunk 5]
```diff
@@ -68,6 +68,27 @@ public final class ATTNSDK: NSObject {
...
+        get { ATTNLogBuffer.shared.isCapturing }
+        set { ATTNLogBuffer.shared.isCapturing = newValue }
+    }
+
+    /// Recent SDK log entries from the in-process ring buffer. Returns an empty
+    /// array when log capture is disabled (the default).
+    /// - Parameter since: Optional cutoff; entries older than this are filtered out.
+    public static func recentLogs(since: Date? = nil) -> [ATTNLogEntry] {
+        ATTNLogBuffer.shared.snapshot(since: since)
+    }
```
Comment (adelag-attentive) [inline Sources/Public/SDK/ATTNSDK.swift]: **Missing `@objc` on new public APIs**

Per the SDK's coding standards, all public API surface should have `@objc` annotations for Objective-C consumers. `isLogCaptureEnabled` (Bool) and `recentLogs(since:)` (returns `[ATTNLogEntry]` where `ATTNLogEntry` is an NSObject subclass) are both ObjC-bridgeable and should get `@objc`.

`logStream` can't be bridged (AsyncStream isn't representable in ObjC), so that one is fine without it.

**Review by adelag-attentive — COMMENTED** [review]
Two review comments — one on interpolation cost, one on @objc bridging.

**Review by adelag-attentive — COMMENTED** [review]
## Review Comments

### 1. `@autoclosure` + `@escaping` in `ATTNLogger` — still evaluates on the `else` branch (Medium)

```swift
func debug(_ message: @autoclosure @escaping () -> ATTNLogMessage) {
    if ATTNLogBuffer.shared.isCapturing {
        let rendered = message().rendered
        osLogger.debug("\(rendered, privacy: .public)")
        ...
    } else {
        osLogger.debug("\(message().rendered, privacy: .public)")  // ← evaluates here too
    }
}
```

The `else` branch still eagerly evaluates `message()` to produce the string for `os.Logger`. The comment says "os.Logger keeps its native gating" but that's not quite right — `os.Logger.debug` would normally not evaluate its `OSLogMessage` argument on release builds (it's compiler-optimized away). Here, you're pre-rendering into a `String` and passing it as `"\(rendered, privacy: .public)"`, which forces evaluation regardless of whether unified logging would have dropped it.

**The net effect:** when `isCapturing = false` and on a release build, every `Loggers.*.debug(...)` call still allocates a `String` via `ATTNLogMessage` interpolation. This is a regression from the original `os.Logger` direct usage where `.debug` was truly zero-cost on release.

**Suggestion:** Consider a fast-path: if `!isCapturing` AND the level is `.debug`, skip entirely (don't even call `osLogger.debug`). Or, only forward to `osLogger` using its native string interpolation (not pre-rendered strings) when capture is off — though this is hard to do generically.

This is the one thing I'd want addressed before merge, since the SDK has ~178 `.debug` call sites that now all allocate on every invocation in production.

### 2. Overlay window may steal key window status (Medium — potential UI issue)

```swift
window.windowLevel = .alert + 1
window.isHidden = false
```

The overlay window is made visible immediately during `scene(_:willConnectTo:)` — the same method where the app's main window calls `makeKeyAndVisible()`. Since the overlay window is created *after* `makeKeyAndVisible()`, UIKit may promote it to the key window. If it does:
- Text fields in the app beneath might not receive keyboard focus properly
- `UIApplication.shared.keyWindow` (deprecated but still used by some code/third-party libraries) could return the overlay window
- Any SDK code or third-party library that uses `keyWindow` to present alerts, action sheets, or web views could attach them to the overlay window instead of the app window

**Fix:** Override `canBecomeKey` on `OverlayWindow` to return `false`:

```swift
override var canBecomeKey: Bool { false }
```

This ensures the overlay never steals key window status regardless of window level or creation order.

### 3. No bounds checking on drag — button/panel can be dragged off-screen with no recovery (Medium — UX issue)

Both the LOGS button and the panel use an unclamped drag handler:

```swift
@objc private func handleDrag(_ gesture: UIPanGestureRecognizer) {
    guard let target = gesture.view, let parent = target.superview else { return }
    let translation = gesture.translation(in: parent)
    target.center = CGPoint(x: target.center.x + translation.x, y: target.center.y + translation.y)
    gesture.setTranslation(.zero, in: parent)
}
```

If the user drags the LOGS button fully off-screen, there's no way to get it back — no reset mechanism exists. On device rotation, the button stays at its dragged position (`didPositionToggle = true` after first layout), so if it was off-screen in portrait it stays off-screen in landscape.

**Suggestion:** Clamp the center so at least 20px of the view remains visible within the parent's bounds. Or add a double-tap-to-reset gesture on the button.

**Comment by adelag-attentive** [comment]
shipping this to unblock inbox e2e testing -
can you add a toggle button (if it's not already there just in case) and then go ahead and ship that PR?

**Review by adelag-attentive — DISMISSED** [review]
per slack thread https://attentivemobile.slack.com/archives/C09PRH450MC/p1784663381587279?thread_ts=1784662494.358519&cid=C09PRH450MC

**Comment by usharif-attentive** [comment]
Rebased onto `develop` and retargeted the PR base from `main` to `develop` (the two branches had diverged; the diff is now just this feature's commits). Status of review feedback:

**Already addressed in earlier follow-up commits:**
- **Eager interpolation on release builds** — `defer log interpolation when buffer capture is off`: each level method takes the message as an `@autoclosure` and, when the buffer isn't capturing, forwards it into `os.Logger`'s own lazy interpolation, so disabled `.debug` calls cost nothing. Capture is also now opt-in via `ATTNSDK.isLogCaptureEnabled` (off by default; Bonni enables it at launch).
- **Unbounded `AsyncStream` buffering** — `stream()` now uses `.bufferingNewest(capacity)`, so a lagging subscriber drops oldest pending entries instead of growing without bound.
- **`ATTNLogEntry.level` stringly-typed** — added `logLevel: Level?` typed accessor and documented the invariant on `level`.
- **Missing `@objc`** — `isLogCaptureEnabled` and `recentLogs(since:)` are `@objc`; `logStream` can't bridge.
- **Panel trim/lineCount bookkeeping** — panel now keeps a `lines: [String]` source of truth with `renderedCount`/`pendingSplice`, splices leading lines in place instead of rebuilding `textView.text`, and no longer undercounts multi-line flushes.
- **Ring buffer test coverage** — `ATTNLogBufferTests` covers the eviction boundary and subscriber add/terminate lifecycle.

**New in this push:**
- **Key window stealing** — `OverlayWindow` overrides `canBecomeKey` to return `false`.
- **Drag off-screen with no recovery** — drags on the LOGS button and panel are clamped so at least 24pt stays on-screen, and positions re-clamp on rotation.
- Moved the debug-log API into an `ATTNSDK` extension — the rebase pushed the class body past SwiftLint's 700-line `type_body_length` error threshold.

Verified: full test suite passes on iPhone 17 sim (228 passed, 0 failed) and Bonni builds cleanly.

**Comment by usharif-attentive** [comment]
Added a **"Show SDK Log Overlay"** switch to Bonni's Settings screen (`4372beb`) — the overlay chrome (LOGS button included) can now be hidden entirely. The preference persists in `UserDefaults`, applies live, and defaults to **visible** so QA/TestFlight behavior is unchanged. Log capture keeps running while hidden, so re-enabling the overlay still shows full history. Verified both states in the simulator.
